data_structures/src/chain/priority.rs

```diff
@@ -0,0 +1,1337 @@
+use std::{cmp, collections::VecDeque, convert, fmt, ops, time::Duration};
+
+use itertools::Itertools;
+
+use failure::Fail;
+use ordered_float::OrderedFloat;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use witnet_util::timestamp::seconds_to_human_string;
+
+use crate::{
+    transaction::Transaction,
+    types::visitor::{StatefulVisitor, Visitor},
+    wit::Wit,
+};
+use std::num::TryFromIntError;
+
+// Assuming no missing epochs, this will keep track of priority used by transactions in the last 24
+// hours. This is rounded up to the closest `2 ^ n - 1` because `PriorityEngine` uses a `VecDeque`
+// under the hood.
+const DEFAULT_QUEUE_CAPACITY_EPOCHS: usize = 2047;
+// The minimum number of epochs that we need to track before estimating transaction priority
+const MINIMUM_TRACKED_EPOCHS: u32 = 20;
+
+/// Keeps track of fees being paid by transactions included in recent blocks, and provides methods
+/// for estimating sensible priority values for future transactions.
+///
+/// This supports _value transfer transactions_ (VTTs) as well as _data requests_ (DRs).
+///
+/// All across this module, fees are always expressed in their relative form (nanowits per weight
+/// unit), aka "transaction priority".
+#[derive(Clone, Eq, PartialEq)]
+pub struct PriorityEngine {
+    /// Soft-capped capacity for the inner priorities queue.
+    capacity: usize,
+    /// Queue for storing fees info for recent transactions.
+    priorities: VecDeque<Priorities>,
+}
+
+impl PriorityEngine {
+    /// Retrieve the inner fee information as a vector.
+    pub fn as_vec(&self) -> Vec<Priorities> {
+        self.priorities.iter().rev().cloned().collect_vec()
+    }
+
+    /// Provide suggestions for sensible transaction priority values, together with their expected
+    /// time-to-block in epochs.
+    ///
+    /// This is only a first approach to an estimation algorithm. There is abundant prior art about
+    /// fee estimation in other blockchains. We might revisit this once we collect more insights
+    /// about our fees market and user feedback.
+    pub fn estimate_priority(
+        &self,
+        seconds_per_epoch: Duration,
+    ) -> Result<PrioritiesEstimate, PriorityError> {
+        // Short-circuit if there are too few tracked epochs for an accurate estimation.
+        let len = u32::try_from(self.priorities.len())?;
```

Comment:
This try_from can only fail if there are more than u32::MAX elements, and in that case the len will always be greater than MINIMUM_TRACKED_EPOCHS, so we shouldn't need to return an error.

```suggestion
        let len = u32::try_from(self.priorities.len()).unwrap_or(u32::MAX);
```